pkg/checker/checker.go

```diff
@@ -6,49 +6,113 @@ package checker
 
 import "fmt"
 
-// Checker is the type of checking functions
-type Checker func() error
+// ResultOK is the string to appear in CheckResult.Result if the check was successful
+const ResultOK = "OK"
 
-// Remediator is the type of remediation functions
-type Remediator func() error
+//ResultError is the string to appear in CheckResult.Result if the check failed
+const ResultError = "ERROR"
 
-// Check is a type that implements a netboot check
+// CheckFun is the type of checking functions
+type CheckFun func(CheckArgs) error
+
+// CheckArgs is the argument map to be passed to CheckFuns
+type CheckArgs map[string]string
+
+// A Check describes a call to a CheckFun with possible remediations
 type Check struct {
-	Name        string
-	Run         Checker
-	Remediate   Remediator
-	StopOnError bool
+	Description string `json:"description"`
+
+	CheckerFuncName string    `json:"checkFun"`
+	CheckerFuncArgs CheckArgs `json:"checkArgs"`
+
+	Remediations  []Check `json:"remediations"`
+	StopOnFailure bool    `json:"stopOnFailure"`
+}
+
+// A CheckResult describes the result of Running a Check
+type CheckResult struct {
+	Description string `json:"description"`
+
+	CheckerFuncName string    `json:"checkFun"`
+	CheckerFuncArgs CheckArgs `json:"checkArgs"`
+
+	Result             string        `json:"result"`
+	Error              string        `json:"error"`
+	RemediationResults []CheckResult `json:"remediationResults,omitempty"`
+	StoppedOnFailure   bool          `json:"-"`
 }
 
-// Run runs the checks and remediations from a check list, in order, and prints the
-// check and remediation status.
-func Run(checklist []Check) error {
-	for idx, check := range checklist {
-		fmt.Printf(green("#%d", idx+1)+" Running check '%s'.. ", check.Name)
-		if checkErr := check.Run(); checkErr != nil {
-			fmt.Println(red("failed: %v", checkErr))
-			if check.Remediate != nil {
-				fmt.Println(yellow("  -> running remediation"))
-				if remErr := check.Remediate(); remErr != nil {
-					fmt.Printf(red("     Remediation for '%s' failed: %v\n", check.Name, remErr))
-					if check.StopOnError {
-						fmt.Println("Exiting")
-						return remErr
-					}
-				} else {
-					fmt.Printf("     Remediation for '%s' succeeded\n", check.Name)
-				}
-			} else {
-				msg := fmt.Sprintf(" -> no remediation found for %s", check.Name)
-				if check.StopOnError {
-					fmt.Println(yellow(msg + ", stop on error requested. Exiting."))
-					return checkErr
-				}
-				fmt.Println(yellow(msg + ", skipping."))
+// Run a Check
+func (check *Check) Run() CheckResult {
+	return check.run(0)
+}
+
+func (check *Check) run(lvl int) CheckResult {
+	result := CheckResult{
+		Description:     check.Description,
```

Comment:
I'm not sure it makes sense to me for this function to return 3 things from the check itself... The caller already knows those things?

I don't see why Description and CheckerFunc* have to be a part of this struct

Reply:
These were kept to give user-friendly feedback to the user running the check (the final output of the cli util is a list of `CheckResult`, serialised to JSON).

To remove these fields and still provide this information to the user, we could do some sort of recursive zip-like operation between each `Check` and its corresponding `CheckResult` to reduce the amount of duplicate state (each `Check` can have 0 or more remediations, which are `Check`s that run if the original check fails to give more troubleshooting information to the user).

Alternatively, we could store a pointer to the corresponding `Check` and implement custom marshalling code to show these fields to the user, or introduce a new data structure just for printing.

What I dislike about the solutions above is that we'd remove the 1-1 relationship we have between the JSON output from the cli and the `CheckResult` struct, which makes the code simpler to maintain.